packages/data-api/src/AnalyticsFetchQuery.js

```diff
@@ -3,119 +3,130 @@
  */
 
-import { DataFetchQuery, ANSWER_SPECIFIC_FIELDS } from './DataFetchQuery';
+import { DataFetchQuery } from './DataFetchQuery';
+
+const VALUE_AGGREGATION_FUNCTIONS = {
+  SUM: 'sum(value::numeric)::text',
+  MOST_RECENT: 'most_recent(value, date)',
+};
 
 const AGGREGATION_SWITCHES = {
   FINAL_EACH_DAY: {
     groupByPeriodField: 'day_period',
-    getLatestPerPeriod: true,
+    aggregationFunction: VALUE_AGGREGATION_FUNCTIONS.MOST_RECENT,
   },
   FINAL_EACH_WEEK: {
     groupByPeriodField: 'week_period',
-    getLatestPerPeriod: true,
+    aggregationFunction: VALUE_AGGREGATION_FUNCTIONS.MOST_RECENT,
+  },
+  SUM_EACH_WEEK: {
+    groupByPeriodField: 'week_period',
+    aggregationFunction: VALUE_AGGREGATION_FUNCTIONS.SUM,
   },
   FINAL_EACH_MONTH: {
     groupByPeriodField: 'month_period',
-    getLatestPerPeriod: true,
+    aggregationFunction: VALUE_AGGREGATION_FUNCTIONS.MOST_RECENT,
   },
   FINAL_EACH_YEAR: {
     groupByPeriodField: 'year_period',
-    getLatestPerPeriod: true,
+    aggregationFunction: VALUE_AGGREGATION_FUNCTIONS.MOST_RECENT,
   },
   MOST_RECENT: {
-    getLatestPerPeriod: true,
+    aggregationFunction: VALUE_AGGREGATION_FUNCTIONS.MOST_RECENT,
   },
   MOST_RECENT_PER_ORG_GROUP: {
-    getLatestPerPeriod: true,
+    aggregationFunction: VALUE_AGGREGATION_FUNCTIONS.MOST_RECENT,
     aggregateEntities: true,
   },
   SUM_PER_ORG_GROUP: {
-    sum: true,
+    aggregationFunction: VALUE_AGGREGATION_FUNCTIONS.SUM,
     aggregateEntities: true,
   },
   SUM_PER_PERIOD_PER_ORG_GROUP: {
-    sum: true,
+    aggregationFunction: VALUE_AGGREGATION_FUNCTIONS.SUM,
     aggregateEntities: true,
-    groupByPeriodField: 'day_period', // can assume first internal aggregation period type is daily
   },
 };
 
 export class AnalyticsFetchQuery extends DataFetchQuery {
   constructor(database, options) {
     super(database, options);
 
-    const firstAggregation = options.aggregations?.[0];
-    this.aggregation = { type: firstAggregation?.type };
-    this.aggregation.switches = AGGREGATION_SWITCHES[firstAggregation?.type]; // add internal switches
-    this.aggregation.config = firstAggregation?.config; // add external config, supplied by client
-    this.isAggregating = !!this.aggregation.switches;
+    this.aggregations = [];
+    const originalAggregations = [...options.aggregations];
+    for (let i = 0; i < originalAggregations.length; i++) {
+      const aggregation = originalAggregations[i];
+      if (!AGGREGATION_SWITCHES[aggregation?.type]) {
+        break; // We only support chaining aggregations if all aggregations are supported type
+      }
+      const dbAggregation = { type: aggregation?.type };
+      dbAggregation.switches = AGGREGATION_SWITCHES[aggregation?.type]; // add internal switches
+      dbAggregation.config = aggregation?.config; // add external config, supplied by client
+      dbAggregation.stackId = i + 1;
+      this.aggregations.push(dbAggregation);
+      options.aggregations.shift();
```

Comment:
This is new, we are now `unshifting` the aggregation from the options, basically meaning that `Aggregator` now only runs the remaining aggregations. Couple things of note here:
- We now must _**fully**_ support the aggregation in the database level! Aggregator will no longer pick up the slack for us if we do a half assed job of it 😝 
- Mutating the aggregations list here is very 'un functional' of me, and has a huge risk for bugs being introduced around it... I think if we moved to using a proper `Queue` implementation for the aggregation list, and we could `enQueue` and `deQueue` cleanly then I wouldn't have any issues... how does the reviewer feel about me giving that a shot?

Reply:
Heh, yes it's a bit of a worry that we have to do things properly! ;-) 

I think the interface could be more encapsulated than that, so that a) we're not nastily mutating this array, but also b) we could plausibly transition to an http interface rather than a functional one in future (i.e. there could be a split somewhere between aggregator and Tupaia's data-api). This would allow having our data-api be a completely different system, and also mean we could potentially make use of other data services' inbuilt aggregation, and have them flag it back to the aggregator in a similar way.

The response from data-broker back to the aggregator is already an object, so there's room to put such information... what do you think?

![image](https://user-images.githubusercontent.com/1274422/126935989-11ec240b-e0b9-4957-ab7b-8d7ee220317f.png)

Reply:
> This would allow having our data-api be a completely different system, and also mean we could potentially make use of other data services' inbuilt aggregation, and have them flag it back to the aggregator in a similar way.

I'm really glad to hear your say this! I've actually been thinking along the lines that it might be nice if our internal datasources (TupaiaDataApi and Indicators) could be run as separate services!

I like that suggestion 👍